File: problem/regexredux/solution/lua/1.lua
-- The Computer Language Benchmarks Game
-- https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
--
-- Contributed by Jeremy Zerfas.
--
-- WARNING: I normally do my programming in other languages. This may not be an
-- optimal Lua program. Please contribute a better program if you can make a
-- better program.

-- We'll be using the PCRE2 flavor of Lrexlib for our regular expression needs
-- instead of using Lua's built in pattern matching engine. Lua's pattern
-- matching engine is a bit basic and lacks some functionality of other regular
-- expression engines, in particular it doesn't support alternation (e.g. you
-- can't use a pattern like "cat|dog" to search for "cat" or "dog"). It is
-- possible to work around this issue by splitting up the regular expressions
-- into multiple smaller patterns but this could possibly be considered as using
-- a different algorithm and it can result in a faster performing program.
-- Additionally PCRE2 is significantly faster than Lua's pattern matching engine
-- anyway.
Lrexlib_PCRE2=require("rex_pcre2")


-- Read in input from stdin and also get the input_Length.
input=io.read("*all")
input_Length=#input


-- Find all sequence descriptions and new lines in input, replace them with
-- empty strings, and store the result in the sequences string.
regex=Lrexlib_PCRE2.new(">.*\\n|\\n")
regex:jit_compile()
sequences=Lrexlib_PCRE2.gsub(input, regex, "")
sequences_Length=#sequences


-- Iterate through all the count patterns and output the results for each one.
for _, pattern in ipairs({
    "agggtaaa|tttaccct",
    "[cgt]gggtaaa|tttaccc[acg]",
    "a[act]ggtaaa|tttacc[agt]t",
    "ag[act]gtaaa|tttac[agt]ct",
    "agg[act]taaa|ttta[agt]cct",
    "aggg[acg]aaa|ttt[cgt]ccct",
    "agggt[cgt]aa|tt[acg]accct",
    "agggta[cgt]a|t[acg]taccct",
    "agggtaa[cgt]|[acg]ttaccct",
  }) do
        regex=Lrexlib_PCRE2.new(pattern)
        regex:jit_compile()
        io.write(string.format("%s %d\n", pattern, Lrexlib_PCRE2.count(sequences
,
          regex)))
end


-- Copy sequences to postreplace_Sequences and then iterate through all the
-- replacement pairs applying all the replacements to postreplace_Sequences.
postreplace_Sequences=sequences
for _, pattern_Replacement_Pair in ipairs({
    {"tHa[Nt]", "<4>"},
    {"aND|caN|Ha[DS]|WaS", "<3>"},
    {"a[NSt]|BY", "<2>"},
    {"<[^>]*>", "|"},
    {"\\|[^|][^|]*\\|", "-"},
  }) do
        regex=Lrexlib_PCRE2.new(pattern_Replacement_Pair[1])
        regex:jit_compile()
        postreplace_Sequences=Lrexlib_PCRE2.gsub(postreplace_Sequences, regex,
          pattern_Replacement_Pair[2])
end


-- Print the size of the original input, the size of the input without the
-- sequence descriptions & new lines, and the size after having made all the
-- replacements.
io.write(string.format("\n%d\n%d\n%d\n", input_Length, sequences_Length,
  #postreplace_Sequences))

File: problem/regexredux/solution/php/1.php
<?php
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   regex-dna program contributed by Danny Sauer
   modified by Josh Goldfoot
   modified by Sergey Khripunov
   modified by Craig Russell
   converted from regex-dna program
*/

$tok = ftok(__FILE__, chr(time() & 255));
$queue = msg_get_queue($tok);

$variants = array(
    ʼagggtaaa|tttaccctʼ,
    ʼ[cgt]gggtaaa|tttaccc[acg]ʼ,
    ʼa[act]ggtaaa|tttacc[agt]tʼ,
    ʼag[act]gtaaa|tttac[agt]ctʼ,
    ʼagg[act]taaa|ttta[agt]cctʼ,
    ʼaggg[acg]aaa|ttt[cgt]ccctʼ,
    ʼagggt[cgt]aa|tt[acg]accctʼ,
    ʼagggta[cgt]a|t[acg]taccctʼ,
    ʼagggtaa[cgt]|[acg]ttaccctʼ,
);

// IUB replacement parallel arrays
$IUB = array();                 $IUBnew = array();
$IUB[]=ʼ/tHa[Nt]/Sʼ;            $IUBnew[]=ʼ<4>ʼ;
$IUB[]=ʼ/aND|caN|Ha[DS]|WaS/Sʼ; $IUBnew[]=ʼ<3>ʼ;
$IUB[]=ʼ/a[NSt]|BY/Sʼ;          $IUBnew[]=ʼ<2>ʼ;
$IUB[]=ʼ/<[^>]*>/Sʼ;            $IUBnew[]=ʼ|ʼ;
$IUB[]=ʼ/\\|[^|][^|]*\\|/Sʼ;    $IUBnew[]=ʼ-ʼ;


// read in file
$contents = file_get_contents(ʼphp://stdinʼ);
$initialLength = strlen($contents);

// remove things
$contents = preg_replace(ʼ/^>.*$|\n/mSʼ, ʼʼ, $contents);
$codeLength = strlen($contents);

// do regexp counts
$messages = array_flip($variants);
$workers = $results = array();
foreach ($variants as $key => $regex){
   if($key == 0 || $key == 2 || $key == 4 || $key == 6) {
      $pid = pcntl_fork();
      if($pid) $workers[] = $pid;
   }
   if($pid && $key > 7) {
      $messages[$regex] =
         preg_match_all(ʼ/ʼ . $regex . ʼ/iSʼ, $contents, $discard);
   }
   else if(!$pid) {
      $results[] = $regex . ʼ,ʼ .
         preg_match_all(ʼ/ʼ . $regex . ʼ/iSʼ, $contents, $discard);
      if($key == 1 || $key == 3 || $key == 5 || $key == 7) {
         msg_send($queue, 2, implode(ʼ;ʼ, $results), false, false, $errno);
         exit;
          }
   }
}

// receive and output the counts
pcntl_wait($status);
foreach($workers as $worker) {
   msg_receive($queue, 2, $msgtype, 4096, $message, false);
   $message = explode(ʼ;ʼ, $message, 3);
   foreach($message as $line) {
      $tmp = explode(ʼ,ʼ, $line, 2);
      $messages[$tmp[0]] = $tmp[1];
   }
}
foreach($messages as $regex => $count) {
   echo $regex, ʼ ʼ, $count, "\n";
}

// do replacements
$contents = preg_replace($IUB, $IUBnew, $contents);

echo "\n",
      $initialLength, "\n",
      $codeLength, "\n",
      strlen($contents), "\n";

